### PR TITLE
Add `task_acks_late` configuration to Celery Executor

### LINKED ISSUE #16163: Confusing log for long running tasks: "dependency 'Task Instance Not Running' FAILED: Task is in the running state"
**Apache Airflow version**: 1.10.* / 2.0.* / 2.1.*

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): Any

**Environment**:

- **Cloud provider or hardware configuration**: Any
- **OS** (e.g. from /etc/os-release): Any
- **Kernel** (e.g. `uname -a`): Any
- **Install tools**: Any
- **Others**: N/A

**What happened**:

This line in the TaskInstance log is very misleading. It seems to happen for tasks that take longer than one hour. When users are waiting for tasks to finish and see this in the log, they often get confused. They may think something is wrong with their task or with Airflow. In fact, this line is harmless. It's simply saying "the TaskInstance is already running so it cannot be run again". 

```
{taskinstance.py:874} INFO - Dependencies not met for <TaskInstance: ... [running]>, dependency 'Task Instance Not Running' FAILED: Task is in the running state
{taskinstance.py:874} INFO - Dependencies not met for <TaskInstance: ... [running]>, dependency 'Task Instance State' FAILED: Task is in the 'running' state which is not a valid state for execution. The task must be cleared in order to be run.
```

**What you expected to happen**:

The confusion is unnecessary. This line should be silenced in the log. Or it should log something clearer.

**How to reproduce it**:

Any task that takes more than an hour to run has this line in the log.

### PR DESCRIPTION
**Pull request by ninsbl**
Celery [task_acks_late](https://docs.celeryq.dev/en/stable/userguide/configuration.html#task-acks-late) is currently hard-coded to be deactivated. However, this causes issues when tasks exceed the `visibility_timeout`.

This PR, adds a `task_acks_late` parameter to the Celery configuration. Default values remain unchanged, but with the new config options users can at least try, if this setting solves there issue if they encounter the `'Task Instance Not Running' FAILED: Task is in the running state` error message.

It is an enhanced re-implementation of #31829 and adresses #16163.

closes #16163
related: #31829

This is not covered by tests (yet). I can add tests but would be happy to receive pointers to e.g. existing test for environment settings, that I could adjust.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by dirrao — CHANGES_REQUESTED** [review]
You can through the following page for writing and running the tests.
https://github.com/apache/airflow/blob/main/contributing-docs/09_testing.rst

File: airflow/providers/celery/executors/default_celery.py
Comment (hussein-awala) [inline airflow/providers/celery/executors/default_celery.py]: We need to keep the default value True
```suggestion
    "task_acks_late":  conf.getboolean("celery", "task_acks_late", fallback=True),
```

File: airflow/providers/celery/provider.yaml
Comment (hussein-awala) [inline airflow/providers/celery/provider.yaml]:
```suggestion
        default: "True"
```

**Review by hussein-awala — COMMENTED** [review]
Could you add a test?

**Comment by hussein-awala** [comment]
I just checked the conversation in the original issue, and personally I prefer to keep the default value to True and waits for community feedback before deciding if we need to change it to False. Let's wait for others reviews.

**Comment by ninsbl** [comment]
> I just checked the conversation in the original issue, and personally I prefer to keep the default value to True and waits for community feedback before deciding if we need to change it to False. Let's wait for others reviews.

Sure, False as default was a glitch. It makes a lot of sense to keep the current default.

I am happy to add tests. What are optional and what are required tests? E.g. I have no experience with Kubernetes and Helm charts. So owuld for example be a unit test that checks if the variable is set properly be enough? With the Kubernetes stuff I may need assistance to get that merge-ready soonish...

**Comment by hussein-awala** [comment]
> I am happy to add tests. What are optional and what are required tests? E.g. I have no experience with Kubernetes and Helm charts. So owuld for example be a unit test that checks if the variable is set properly be enough? With the Kubernetes stuff I may need assistance to get that merge-ready soonish...

You can add a small test to check if the conf is taken into account by Airflow, you can use this test as a reference:
https://github.com/apache/airflow/blob/e2393ee6dd3a927e753e6375621af07aa0c734dc/tests/providers/celery/executors/test_celery_executor.py#L352-L360

**Comment by ninsbl** [comment]
Tried to add a basic unittest in [b9be045](https://github.com/apache/airflow/pull/37066/commits/b9be0459c7f3a6080f3883d43d31e33e1317e0db). Checked it with breeze and it seems to work...

File: tests/providers/celery/executors/test_celery_executor.py
Comment (hussein-awala) [inline tests/providers/celery/executors/test_celery_executor.py]: True is the default value, you need to test with False, you can also create a parametrized test to test with both values.

Comment (ninsbl) [reply tests/providers/celery/executors/test_celery_executor.py]: > True is the default value, you need to test with False, you can also create a parametrized test to test with both values.

Right, sorry. See: [4a847ec](https://github.com/apache/airflow/pull/37066/commits/4a847ecb1863b5035ac5d09e604ff12493a45e5e)

**Review by hussein-awala — APPROVED** [review]
LGTM

File: airflow/providers/celery/provider.yaml
Comment (o-nikolas) [inline airflow/providers/celery/provider.yaml]: Proposed rewording:

```suggestion
          If an Airflow task's execution time exceeds the visibility_timeout, Celery will re-assign the task to a Celery worker, even if the
          original task is still running successfully. The new task instance then runs
          concurrently with the original task and the Airflow UI and logs only show an error message:
          'Task Instance Not Running' FAILED: Task is in the running state'
          Setting task_acks_late to True will force Celery to wait until a task is finished before a
          new task instance is assigned. This effectively overrides the visibility timeout.
          See also:
          https://docs.celeryq.dev/en/stable/reference/celery.app.task.html#celery.app.task.Task.acks_late
```

File: airflow/providers/celery/provider.yaml
Comment (o-nikolas) [inline airflow/providers/celery/provider.yaml]: Why no version specified?

File: airflow/providers/celery/provider.yaml
Comment (o-nikolas) [inline airflow/providers/celery/provider.yaml]:
```suggestion
        example: "True"
```

File: airflow/providers/celery/provider.yaml
Comment (ninsbl) [reply airflow/providers/celery/provider.yaml]: Copy and paste from another section, and the other examples I looked at did not have a version specified. Any suggestions?

**Comment by ninsbl** [comment]
BTW.: When running tests with breeze (on WSL2) I get the following error message:
```
23.16 INFO: pip is looking at multiple versions of apache-airflow[amazon] to determine which version is compatible with other requirements. This could take a while.
23.31 ERROR: Cannot install apache-airflow[amazon]==2.9.0.dev0 because these package versions have conflicting dependencies.
23.31
23.31 The conflict is caused by:
23.31     apache-airflow[amazon] 2.9.0.dev0 depends on moto<5.0.0 and >=4.2.12; extra == "amazon"
23.31     The user requested (constraint) moto==5.0.0
23.31
23.31 To fix this you could try to:
23.31 1. loosen the range of package versions you've specified
23.31 2. remove package versions to allow pip attempt to solve the dependency conflict
23.31
23.31 ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
------
Dockerfile.ci:1190
--------------------
 1188 |     # But in cron job we will install latest versions matching pyproject.toml to see if there is no breaking change
 1189 |     # and push the constraints if everything is successful
 1190 | >>> RUN bash /scripts/docker/install_airflow.sh
 1191 |
 1192 |     COPY --from=scripts entrypoint_ci.sh /entrypoint
--------------------
ERROR: failed to solve: process "/bin/bash -o pipefail -o errexit -o nounset -o nolog -c bash /scripts/docker/install_airflow.sh" did not complete successfully: exit code: 1
```
This does not prevent unittests from being executed though...

**Comment by potiuk** [comment]
> BTW.: When running tests with breeze (on WSL2) I get the following error message:
> 
> This does not prevent unittests from being executed though...

Yes. Yu need to rebase and rerun it to fix it. Breeze /tests works for sure in latest main. Alternatively (but this is something breeze will always print as warning in this case) there are times before merging conflicting changes to main and refreshing constraints where you need to run build with `--upgrade-to-newer-dependencies`.

Actually I think I wil make a small change to handle this case better because we already have feature to retry build with `--upgrade-to-newer-dependencies` when it fails with constraints , so I can flip the flag to run this mode locally by default

**Comment by potiuk** [comment]
I rebased this one and added PR to make future local build work in case we have not managed to refresh the constraints yet https://github.com/apache/airflow/pull/37116 - that will take longer but will finally succeed and should close the loophole here.

**Comment by ninsbl** [comment]
> Yes. Yu need to rebase and rerun it to fix it. Breeze /tests works for sure in latest main.

Thanks, @potiuk , I can confirm that breeze now builds and runs fine after rebase...

**Comment by ninsbl** [comment]
Anything I can do to move this forward?

**Comment by ninsbl** [comment]
Hmm, it seems the spell checking does ot like the new key name:

(`
Incorrect Spelling: 'acks'
Line with Error: 'task_acks_late'
`)...

**Comment by ninsbl** [comment]
Hope adding the exception for `acks` in [8419527](https://github.com/apache/airflow/pull/37066/commits/841952715dcbe36fb21e93fac1d0e5d588eccbae) was correct resolution for the failing spell check test ... ?

File: airflow/providers/celery/provider.yaml
Comment (ninsbl) [reply airflow/providers/celery/provider.yaml]: Should I change anything? Or can this conversation be considered resolved?

Comment (potiuk) [reply airflow/providers/celery/provider.yaml]: Yes. add 2.9.0 - this is the first one new feature will appear.

Comment (ninsbl) [reply airflow/providers/celery/provider.yaml]: > Yes. add 2.9.0 - this is the first one new feature will appear.

Thanks, done in b3bf903

**Review by potiuk — APPROVED** [review]

File: airflow/providers/celery/provider.yaml
Comment (hussein-awala) [inline airflow/providers/celery/provider.yaml]:
```suggestion
        version_added: ~
```

**Review by hussein-awala — APPROVED** [review]

File: airflow/providers/celery/provider.yaml
Comment (potiuk) [reply airflow/providers/celery/provider.yaml]: Ah right. It's celery - but in this case it should be next celery provider version

Comment (potiuk) [reply airflow/providers/celery/provider.yaml]: (3.6.0)

File: airflow/providers/celery/provider.yaml
Comment (hussein-awala) [reply airflow/providers/celery/provider.yaml]: I just reverted this change because it will be released with the celery provider and not the core release. Currently, we don't set the `version_added` for the providers' configurations.

File: airflow/providers/celery/provider.yaml
Comment (potiuk) [inline airflow/providers/celery/provider.yaml]:
```suggestion
        version_added: 3.6.0
```

Comment (potiuk) [reply airflow/providers/celery/provider.yaml]: @hussein-awala ?

File: airflow/providers/celery/provider.yaml
Comment (potiuk) [reply airflow/providers/celery/provider.yaml]: Heh... yeah provider. But we actually set the versions (and those are provider's versions).

File: airflow/providers/celery/provider.yaml
Comment (hussein-awala) [reply airflow/providers/celery/provider.yaml]: yes we can use the next minor release for Celery, I just checked the other providers' configurations and found some version_added set in some of them

**Comment by hussein-awala** [comment]
Congrats on your first commit 🎉

**Pull request merged**
